src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,265 @@
+            engine.AddGas(GetPrice(engine.Snapshot));
+            NameState state = new NameState
+            {
+                Owner = owner,
+                Name = name,
+                Description = "",
+                Expiration = (uint)(engine.Snapshot.PersistingBlock.Timestamp / 1000) + OneYear
+            };
+            Mint(engine, state);
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash), new StorageItem(new byte[] { 0 }));
+            return true;
+        }
+
+        [ContractMethod(0, CallFlags.WriteStates)]
+        private uint Renew(ApplicationEngine engine, string name)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            engine.AddGas(GetPrice(engine.Snapshot));
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            NameState state = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Token).Add(hash)).GetInteroperable<NameState>();
+            engine.Snapshot.Storages.Delete(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash));
+            state.Expiration += OneYear;
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash), new StorageItem(new byte[] { 0 }));
+            return state.Expiration;
+        }
+
+        [ContractMethod(0_03000000, CallFlags.WriteStates)]
+        private void SetAdmin(ApplicationEngine engine, string name, UInt160 admin)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            if (!engine.CheckWitnessInternal(admin)) throw new InvalidOperationException();
+            NameState state = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Token).Add(GetKey(Utility.StrictUTF8.GetBytes(name)))).GetInteroperable<NameState>();
+            if (!engine.CheckWitnessInternal(state.Owner)) throw new InvalidOperationException();
+            state.Admin = admin;
+        }
+
+        [ContractMethod(0_30000000, CallFlags.WriteStates)]
+        private void SetRecord(ApplicationEngine engine, string name, RecordType type, string data)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            switch (type)
+            {
+                case RecordType.A:
+                    if (!IPAddress.TryParse(data, out IPAddress address)) throw new FormatException();
+                    if (address.AddressFamily != AddressFamily.InterNetwork) throw new FormatException();
+                    break;
+                case RecordType.CNAME:
```

Comment:
It is better to check whether the original domain name exists.

Reply:
But the original name can be external.

Reply:
what do you mean?

Reply:
An external name, for example, `google.com`.